File: a1_go/core/algorithms.go
package core

func FirstFit(id Id, size Size, blocks BlockList) int {
	for i := 0; i < blocks.Len(); i++ {
		block := blocks.Get(i)
		if block.size >= size {
			return i
		}
	}
	return -1
}

func BestFit(id Id, size Size, blocks BlockList) int {
	best := -1
	for i := 0; i < blocks.Len(); i++ {
		block := blocks.Get(i)
		if block.size >= size {
			if best == -1 || block.size < blocks.Get(best).size {
				best = i
			}
		}
	}
	return best
}

func WorstFit(id Id, size Size, blocks BlockList) int {
	best := -1
	for i := 0; i < blocks.Len(); i++ {
		block := blocks.Get(i)
		if block.size >= size {
			if best == -1 || block.size > blocks.Get(best).size {
				best = i
			}
		}
	}
	return best
}

File: a1/a1py/src/core/algorithms.py
from typing import Callable, Optional
import core
import core.block
from core.block import Size, Block

Algorithm = Callable[[core.block.Size, list[core.block.Block]], Optional[int]]


def first_fit(size: Size, blocks: list[Block]) -> Optional[int]:
    for i, block in enumerate(blocks):
        if block.size >= size:
            return i
    return None


def best_fit(size: Size, blocks: list[Block]) -> Optional[int]:
    best_i: Optional[int] = None
    for i, block in enumerate(blocks):
        if block.size >= size:
            if not best_i or blocks[best_i].size > block.size:
                best_i = i
    return best_i


def worst_fit(size: Size, blocks: list[Block]) -> Optional[int]:
    worst_i: Optional[int] = None
    for i, block in enumerate(blocks):
        if block.size >= size:
            if not worst_i or blocks[worst_i].size < block.size:
                worst_i = i
    return worst_i
